authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/user/UserPolicyProviderFactory.java

```diff
@@ -164,31 +163,34 @@ public void postInit(KeycloakSessionFactory factory) {
                 UserModel removedUser = ((UserRemovedEvent) event).getUser();
                 RealmModel realm = ((UserRemovedEvent) event).getRealm();
                 ResourceServerStore resourceServerStore = storeFactory.getResourceServerStore();
-                realm.getClients().forEach(clientModel -> {
-                    ResourceServer resourceServer = resourceServerStore.findByClient(clientModel.getId());
 
-                    if (resourceServer != null) {
-                        policyStore.findByType(getId(), resourceServer.getId()).forEach(policy -> {
-                            List<String> users = new ArrayList<>();
-
-                            for (String userId : getUsers(policy)) {
-                                if (!userId.equals(removedUser.getId())) {
-                                    users.add(userId);
-                                }
-                            }
-
-                            try {
-                                if (users.isEmpty()) {
-                                    policyStore.delete(policy.getId());
-                                } else {
-                                    policy.getConfig().put("users", JsonSerialization.writeValueAsString(users));
-                                }
-                            } catch (IOException e) {
-                                throw new RuntimeException("Error while synchronizing users with policy [" + policy.getName() + "].", e);
```

Comment:
@hmlnarik, at the same time I rebase onto master, I'm looking at the code Alex did and I'm challenging this findByClients() (new method) thing entirely. First of all, he's not going through the storeFactory which will return the "cached" item first which will then delegate to the JPA provider if necessary. Pattern seen everywhere else.

What I'm guessing here (I'm not sure I have his test data anymore) is that the forEach() call that when through the clientModel objects to do resourceServerStore.findByClient() likely triggered multiple implicit flush making this part very slow with a large number of clients (what we were testing). That would be caused by the use of the "findResourceServerIdByClient" NamedQuery to find the ResourceServerEntity id, to then find the cached object (or load the entity directly through it's PK)

My question here: Since the RESOURCE_SERVER has a unique constraint on CLIENT_ID, why not dropping the ID table and use the CLIENT_ID as the primary key? That would be flagged as the @Id in ResourceServerEntity and efficient Hibernate cache management would now be implicit (no more flush on query).

Any reason not to have a FK as the PK for RESOURCE_SERVER? Also, it seems that a FK is missing on that table.

Reply:
@glavoie I'm uncertain why RESOURCE_SERVER is structured in this way, in fact it rather looks like extension of CLIENT table with new attributes (inheritance in a way). @pedroigor is the original author, he might have a deeper insight why there is a separate ID column.

Re the missing FK - I understand benefits of ignoring the FK (performance) vs defining it (data consistency). Currently it is used inconsistently across DB, so the FK definition or undefinitions will be part of a bigger database schema cleanup that will need to come in future.

Reply:
@glavoie @hmlnarik I can't remember right now a good reason for not using CLIENT_ID as PK. But change it now may introduce more critical changes (this PR by itself is full of them :)) and risks to existing functionality.

Did you try configuring flush mode to COMMIT for that particular query ? It could avoid flushing to database on every single query.

Reply:
@pedroigor, @hmlnarik: The change I want to submit has nothing to do with what is done in this PR (now closed/not merged). It's mostly removal of getByClient() to switch all calls to getId with the same call value and re-wire the id field on ResourceServerEntity to the CLIENT_ID column. The other part of the fix is to migrate the data in the database to use the client_id instead of the other PK.

I was able to test/make the change in about 10 minutes, but I need to clean things up + create a Liquibase changset for the data migration. Planning on doing that tomorrow morning.

I stopped thinking about changing the flush mode a long time ago in Keycloak, when working on performance. Most of the improvements I found in the past were easier to do by working a bit the JPA model. There are just too many locations where it would be necessary to do it and in my opinion, developers would just not think about it when working on new code areas where it could be necessary to do it. You also never know when a transaction could modify data that would need to be seen in a NamedQuery result right away, before commit.

Reply:
@glavoie, please let me know when you are done with your changes. I would like to cherry-pick them and run some tests.

Reply:
Pedro, we are in sync, I just pushed https://github.com/glavoie/keycloak/tree/KEYCLOAK-4858-ResourceServer with the new change I'd like to suggest for the ResourceServer.

Don't be scared by the large amount of file modified, it's mostly to remove the get/setClientId and use get/setId() instead. There's also a Liquibase changeset to propagate the CLIENT_ID to all children tables of RESOURCE_SERVER and re-wire the PK/FKs. 

If you are OK with that single change, I'd open a PR with just this one. Any other change would be different PRs to keep it simple for tracking/merging.

Reply:
@glavoie Thanks. Go ahead and send the PR, please. Will also do some tests on Monday. If you are done with your tests until there, I think we can merge it early next week.

Reply:
@pedroigor, thanks for the quick reply, I'm submitting the PR. Haven't finished my testing and will continue only on Tuesday at work. I'll keep you posted ASAP.